cordapp-cpk/src/main/kotlin/net/corda/plugins/cpb/CpbPlugin.kt

```diff
@@ -0,0 +1,52 @@
+package net.corda.plugins.cpb
+
+import net.corda.plugins.cpk.CORDAPP_CONFIGURATION_NAME
+import net.corda.plugins.cpk.CORDA_CPK_CONFIGURATION_NAME
+import net.corda.plugins.cpk.CordappPlugin
+import org.gradle.api.GradleException
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.artifacts.Configuration
+import org.gradle.api.artifacts.ExternalDependency
+import org.gradle.api.artifacts.ProjectDependency
+import org.gradle.api.tasks.bundling.Jar
+import org.gradle.api.tasks.bundling.ZipEntryCompression
+
+class CpbPlugin : Plugin<Project> {
+
+    companion object{
+        private const val CPB_TASK_NAME = "cpb"
+        private const val CPB_CONFIGURATION_NAME = CPB_TASK_NAME
+        private const val CPB_ARTIFACT_CLASSIFIER = "package"
+        private const val CPB_FILE_EXTENSION = CPB_TASK_NAME
+    }
+
+    override fun apply(project: Project) {
+        project.plugins.findPlugin(CordappPlugin::class.java) ?:
+            throw GradleException("The 'net.corda.plugins.cpb' plugin needs the 'net.corda.plugins.cordapp-cpk' plugin to be applied first")
+
+        val cpbConfiguration : Configuration = project.configurations.create(CPB_CONFIGURATION_NAME)
+        val cordappConfiguration = project.configurations.named(CORDAPP_CONFIGURATION_NAME).get()
+        cpbConfiguration.extendsFrom(cordappConfiguration)
+        project.tasks.register(CPB_TASK_NAME, Jar::class.java) { cpbTask ->
+            cpbTask.archiveClassifier.set(CPB_ARTIFACT_CLASSIFIER)
+            cpbTask.archiveExtension.set(CPB_FILE_EXTENSION)
+            cpbTask.entryCompression = ZipEntryCompression.STORED
+            cpbConfiguration.allDependencies.forEach { dependency ->
+                when(dependency) {
+                    is ExternalDependency -> {
+                        dependency.artifact {
+                            it.name = dependency.name
+                            it.classifier = "cordapp"
+                            it.type = "cpk"
+                        }
+                    }
+                    is ProjectDependency -> {
+                        dependency.targetConfiguration = CORDA_CPK_CONFIGURATION_NAME
+                    }
+                }
+            }
```

Comment:
If I move this bit to line 31, outside of the `register` closure, the assembled CPB contains jar files instead of CPKs.
This makes no sense at all to me, as far as I understand, the content of the `register` closure (except for eventual `doLast`, `doFirst` callbacks) is also executed in the configuration phase and immediately after line 30

Reply:
Lines inside the closure will only be executed if/when the task is configured, whereas lines outside the closure will be executed as soon as the plugin is applied.

Don't forget that you need to use the `allCordapps` configuration instead of `cordapps` in order to collect all transitive CPKs, but this configuration will not be populated immediately. (See `Coniguration.withDependencies()`.)

You should also be aware that "transitive CPK resolution" is not guaranteed to work (unfortunately), which is another reason why we also need to be able to add CPKs to the `cpb` configuration manually.